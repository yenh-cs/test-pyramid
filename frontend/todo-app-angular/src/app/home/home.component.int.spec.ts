import { TestBed, ComponentFixture, waitForAsync, tick, fakeAsync, flush } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { Store, StoreModule } from '@ngrx/store';
import { HomeComponent } from './home.component';
import { TodoListComponent } from '../todo-list/todo-list.component';
import { TodoDialogComponent } from '../todo-dialog/todo-dialog.component';
import { todosReducer } from '../state/todos/todo.reducer';
import { TodoService } from '../services/todo.service';
import { EffectsModule } from '@ngrx/effects';
import { TodoEffects } from '../state/todos/todo.effect';
import { MatDialog, MatDialogModule, MatDialogRef } from '@angular/material/dialog';
import { of } from 'rxjs';
import { MatTabsModule } from '@angular/material/tabs';
import { Todo } from '../models/todo';
import { MatCardModule } from '@angular/material/card';
import { BrowserAnimationsModule, NoopAnimationsModule } from '@angular/platform-browser/animations';
import { TodosApiActions } from '../state/todos/todo.actions';

const todoItems: Todo[] = [
  {
    "title": "Grocery shopping",
    "description": "1. A pack of carrots\n2. Eggs",
    "completed": false
  },
  {
    "title": "Integration Test",
    "description": "10 tests",
    "completed": false
  },
  {
    "title": "Pick up kid from school",
    "description": "",
    "completed": true
  }
];

describe('AppComponent Integration Test', () => {
  let component: HomeComponent;
  let fixture: ComponentFixture<HomeComponent>;
  let todosService: TodoService;
  let store: Store;

  let todosServiceStub: Partial<TodoService> = {
    getTodos: jest.fn().mockReturnValue(of(todoItems)),
    addTodo: jest.fn(),
    updateTodo: jest.fn(),
    deleteTodo: jest.fn(),
  };

  const matDialogRefStub: Partial<MatDialogRef<any>> = {
    close: () => {},
  };

  beforeEach(async () => {
    TestBed.configureTestingModule({
      declarations: [
        HomeComponent,
        TodoListComponent,
        TodoDialogComponent,
      ],
      imports: [
        StoreModule.forRoot({
          todos: todosReducer,
        }),
        EffectsModule.forRoot([
          TodoEffects
        ]),
        MatDialogModule,
        MatTabsModule,
        MatCardModule,
        NoopAnimationsModule,
      ],
      providers: [
        { provide: TodoService, useValue: todosServiceStub },
        { provide: MatDialog, useClass: MatDialog },
        { provide: MatDialogRef, useValue: matDialogRefStub },
      ],
    }).compileComponents();

    todosService = TestBed.inject(TodoService);
    // TestBed.overrideProvider(TodoService, { useValue: { getTodos: () => of(todoItems) } });
    store = TestBed.inject(Store);
    fixture = TestBed.createComponent(HomeComponent);
    component = fixture.debugElement.componentInstance;
    fixture.detectChanges();
  });

  afterEach(() => {
    jest.clearAllMocks();
    TestBed.resetTestingModule();
  });

  it('should create the component', () => {
    expect(component).toBeTruthy();
  });

  it('should display incomplete To-do list', () => {
    const incompletedTodoListItems = fixture.debugElement.queryAll(By.css('mat-card'));
    expect(incompletedTodoListItems.length).toBe(2);
  });

  it('should display complete To-do list', fakeAsync(() => {
    const completedTab = fixture.debugElement.queryAll(By.css('.mdc-tab'))[1];
    completedTab.nativeElement.click();
    fixture.detectChanges();
    setTimeout(() => {}, 0);
    flush();
    const completedTodoListItems = fixture.debugElement.queryAll(By.css('mat-card'));
    const titleElement = completedTodoListItems[2].query(By.css('mat-card-title'));
    expect(titleElement.nativeElement.textContent).toBe('Pick up kid from school');
  }));

  it('should be able complete an incompleted todo item', fakeAsync(() => {
    let firstIncompletedTodoItem = fixture.debugElement.query(By.css('mat-card'));
    expect(firstIncompletedTodoItem.nativeElement.textContent).toContain(todoItems[0].title);

    const completeButton = firstIncompletedTodoItem
                        .query(By.css('mat-card-actions'))
                        .queryAll(By.css('button'))[2];
    completeButton.nativeElement.click();
    fixture.detectChanges();
    firstIncompletedTodoItem = fixture.debugElement.query(By.css('mat-card'));
    expect(firstIncompletedTodoItem.nativeElement.textContent).not.toContain(todoItems[0].title);

    const completedTab = fixture.debugElement.queryAll(By.css('.mdc-tab'))[1];
    completedTab.nativeElement.click();
    fixture.detectChanges();
    setTimeout(() => {}, 0);
    flush();
    const firstCompletedTodoItem = fixture.debugElement.queryAll(By.css('mat-card'))[1];
    expect(firstCompletedTodoItem.nativeElement.textContent).toContain(todoItems[0].title);
  }));

  fit('should be able complete an incompleted todo item', fakeAsync(() => {
    let firstIncompletedTodoItem = fixture.debugElement.query(By.css('mat-card'));
    expect(firstIncompletedTodoItem.nativeElement.textContent).toContain(todoItems[0].title);
  
    const editButton = firstIncompletedTodoItem
                        .query(By.css('mat-card-actions'))
                        .queryAll(By.css('button'))[1];
    editButton.nativeElement.click();
    fixture.detectChanges();
    tick(500); // Wait for a small delay (adjust the duration if needed)
    expect(fixture).toMatchSnapshot();
  }));
});
