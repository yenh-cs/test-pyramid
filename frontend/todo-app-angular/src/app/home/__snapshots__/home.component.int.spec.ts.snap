// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent Integration Test should be able complete an incompleted todo item 1`] = `
<app-home
  allTodos$={[Function Store]}
  completedTodos$={[Function Store]}
  dialog={[Function MatDialog]}
  incompletedTodos$={[Function Store]}
  newTodo={[Function Object]}
  store={[Function Store]}
>
  <div
    class="container"
  >
    <h3>
      All Todos
    </h3>
    <mat-tab-group
      class="mat-mdc-tab-group mat-primary mat-mdc-tab-group-stretch-tabs"
    >
      <mat-tab-header
        class="mat-mdc-tab-header"
      >
        <button
          aria-hidden="true"
          class="mat-ripple mat-mdc-tab-header-pagination mat-mdc-tab-header-pagination-before mat-mdc-tab-header-pagination-disabled"
          disabled=""
          mat-ripple=""
          tabindex="-1"
          type="button"
        >
          <div
            class="mat-mdc-tab-header-pagination-chevron"
          />
        </button>
        <div
          class="mat-mdc-tab-label-container _mat-animation-noopable"
        >
          <div
            class="mat-mdc-tab-list"
            role="tablist"
            style="transform: translateX(0px);"
          >
            <div
              class="mat-mdc-tab-labels"
            >
              <div
                aria-disabled="false"
                aria-posinset="1"
                aria-selected="true"
                aria-setsize="2"
                cdkmonitorelementfocus=""
                class="mdc-tab mat-mdc-tab mat-mdc-focus-indicator mdc-tab--active mdc-tab-indicator--active"
                mattablabelwrapper=""
                role="tab"
                tabindex="0"
              >
                <span
                  class="mdc-tab__ripple"
                />
                <div
                  class="mat-ripple mat-mdc-tab-ripple"
                  mat-ripple=""
                />
                <span
                  class="mdc-tab__content"
                >
                  <span
                    class="mdc-tab__text-label"
                  >
                    Incompleted
                  </span>
                </span>
                <span
                  class="mdc-tab-indicator"
                >
                  <span
                    class="mdc-tab-indicator__content mdc-tab-indicator__content--underline"
                  />
                </span>
              </div>
              <div
                aria-disabled="false"
                aria-posinset="2"
                aria-selected="false"
                aria-setsize="2"
                cdkmonitorelementfocus=""
                class="mdc-tab mat-mdc-tab mat-mdc-focus-indicator"
                mattablabelwrapper=""
                role="tab"
                tabindex="-1"
              >
                <span
                  class="mdc-tab__ripple"
                />
                <div
                  class="mat-ripple mat-mdc-tab-ripple"
                  mat-ripple=""
                />
                <span
                  class="mdc-tab__content"
                >
                  <span
                    class="mdc-tab__text-label"
                  >
                    Completed
                  </span>
                </span>
                <span
                  class="mdc-tab-indicator"
                >
                  <span
                    class="mdc-tab-indicator__content mdc-tab-indicator__content--underline"
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
        <button
          aria-hidden="true"
          class="mat-ripple mat-mdc-tab-header-pagination mat-mdc-tab-header-pagination-after mat-mdc-tab-header-pagination-disabled"
          disabled=""
          mat-ripple=""
          tabindex="-1"
          type="button"
        >
          <div
            class="mat-mdc-tab-header-pagination-chevron"
          />
        </button>
      </mat-tab-header>
      <div
        class="mat-mdc-tab-body-wrapper _mat-animation-noopable"
      >
        <mat-tab-body
          class="mat-mdc-tab-body mat-mdc-tab-body-active"
          role="tabpanel"
        >
          <div
            cdkscrollable=""
            class="mat-mdc-tab-body-content ng-trigger ng-trigger-translateTab"
            style="transform:none;0:transform;"
          >
            <app-todo-list
              style=""
            >
              <mat-card
                class="mat-mdc-card mdc-card todo-card mat-elevation-z10"
              >
                <mat-card-header
                  class="mat-mdc-card-header"
                >
                  <div
                    class="mat-mdc-card-header-text"
                  >
                    <mat-card-title
                      class="mat-mdc-card-title todo-title"
                    >
                      Grocery shopping
                    </mat-card-title>
                  </div>
                </mat-card-header>
                <mat-card-content
                  class="mat-mdc-card-content"
                  style="white-space: pre-line;"
                >
                  <p>
                    1. A pack of carrots
  2. Eggs
                  </p>
                </mat-card-content>
                <mat-card-actions
                  class="mat-mdc-card-actions mdc-card__actions todo-actions"
                >
                  <button
                    color="accent"
                    mat-raised-button=""
                  >
                     Remove 
                  </button>
                  <button
                    color="accent"
                    mat-raised-button=""
                  >
                     Edit 
                  </button>
                  <button
                    color="primary"
                  >
                     Complete 
                  </button>
                </mat-card-actions>
              </mat-card>
              <mat-card
                class="mat-mdc-card mdc-card todo-card mat-elevation-z10"
              >
                <mat-card-header
                  class="mat-mdc-card-header"
                >
                  <div
                    class="mat-mdc-card-header-text"
                  >
                    <mat-card-title
                      class="mat-mdc-card-title todo-title"
                    >
                      Integration Test
                    </mat-card-title>
                  </div>
                </mat-card-header>
                <mat-card-content
                  class="mat-mdc-card-content"
                  style="white-space: pre-line;"
                >
                  <p>
                    10 tests
                  </p>
                </mat-card-content>
                <mat-card-actions
                  class="mat-mdc-card-actions mdc-card__actions todo-actions"
                >
                  <button
                    color="accent"
                    mat-raised-button=""
                  >
                     Remove 
                  </button>
                  <button
                    color="accent"
                    mat-raised-button=""
                  >
                     Edit 
                  </button>
                  <button
                    color="primary"
                  >
                     Complete 
                  </button>
                </mat-card-actions>
              </mat-card>
            </app-todo-list>
          </div>
        </mat-tab-body>
        <mat-tab-body
          class="mat-mdc-tab-body"
          role="tabpanel"
        >
          <div
            cdkscrollable=""
            class="mat-mdc-tab-body-content ng-trigger ng-trigger-translateTab"
            style="transform:translate3d(100%, 0, 0);min-height:1px;visibility:hidden;0:transform;1:min-height;2:visibility;"
          >
          </div>
        </mat-tab-body>
      </div>
    </mat-tab-group>
    <footer>
      <button>
        <img
          class="icon"
          src="assets/add.png"
        />
      </button>
    </footer>
  </div>
</app-home>
`;
